pkg/types/interfacetests/utils.go

```diff
@@ -36,12 +41,91 @@ func runTests[T TestingT[T]](t T, tester BasicTester[T], tests []testcase[T]) {
 	}
 }
 
+// Batch chain write takes a batch call entry and writes it to the chain using the ChainWriter.
+func batchChainWrite[T TestingT[T]](t T, tester ChainComponentsInterfaceTester[T], batchCallEntry BatchCallEntry, mockRun bool) {
+	// This is necessary because the mock helper function requires the entire batchCallEntry rather than an individual testStruct
+	if mockRun {
+		cw := tester.GetChainWriter(t)
+		err := cw.SubmitTransaction(tests.Context(t), AnyContractName, "batchChainWrite", batchCallEntry, "", "", nil, big.NewInt(0))
+		require.NoError(t, err)
+		return
+	}
+	nameToAddress := make(map[string]string)
+	boundContracts := tester.GetBindings(t)
+	for _, bc := range boundContracts {
+		nameToAddress[bc.Name] = bc.Address
+	}
+
+	// For each contract in the batch call entry, submit the read entries to the chain
+	for contractName, contractBatch := range batchCallEntry {
+		require.Contains(t, nameToAddress, contractName)
+		for _, readEntry := range contractBatch {
+			val, isOk := readEntry.ReturnValue.(*TestStruct)
+			if !isOk {
+				require.Fail(t, "expected *TestStruct for contract: %s read: %s, but received %T", contractName, readEntry.Name, readEntry.ReturnValue)
+			}
+			SubmitTransactionToCW(t, tester, MethodSettingStruct, val, types.BoundContract{Name: contractName, Address: nameToAddress[contractName]}, types.Unconfirmed)
+		}
+	}
+}
+
+// SubmitTransactionToCW submits a transaction to the ChainWriter and waits for it to reach the given status.
+func SubmitTransactionToCW[T TestingT[T]](t T, tester ChainComponentsInterfaceTester[T], method string, args any, contract types.BoundContract, status types.TransactionStatus) string {
+	tester.DirtyContracts()
+	txID := uuid.New().String()
+	cw := tester.GetChainWriter(t)
+	err := cw.SubmitTransaction(tests.Context(t), contract.Name, method, args, txID, contract.Address, nil, big.NewInt(0))
+	require.NoError(t, err)
+
+	err = WaitForTransactionStatus(t, tester, txID, status, false)
+	require.NoError(t, err)
+
+	return txID
+}
+
+// WaitForTransactionStatus waits for a transaction to reach the given status.
+func WaitForTransactionStatus[T TestingT[T]](t T, tester ChainComponentsInterfaceTester[T], txID string, status types.TransactionStatus, mockRun bool) error {
+	ctx, cancel := context.WithTimeout(tests.Context(t), 15*time.Minute)
+	defer cancel()
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("transaction %s not finalized within timeout period", txID)
+		case <-ticker.C:
+			if mockRun {
+				tester.GenerateBlocksTillConfidenceLevel(t, "", "", primitives.Finalized)
+				return nil
+			}
```

Comment:
I don't see the EVM interface tester having an impl. for `GenerateBlocksTillConfidenceLevel()`. Does this have an implementation anywhere? If not we should just remove mockRun completely

Reply:
I only have this so that the `fakeChainReader` tests can run [GenerateBlocksTillConfidenceLevel](https://github.com/smartcontractkit/chainlink-common/blob/c7d0f8681a84627d94180fabfec95470d7fca7f5/pkg/loop/internal/relayer/pluginprovider/chainreader/chain_reader_test.go#L569-L579). You're right, this isn't used by the actual implementations but it was necessary for these tests so I wasn't sure how to remove the method without breaking this test.

Reply:
Yeah, I don't see a trivial solution, lets create a ticket just for this

Reply:
[Done](https://smartcontract-it.atlassian.net/jira/software/c/projects/BCI/boards/23?assignee=61a6f090ef18ca00717f84db&quickFilter=398&selectedIssue=BCI-4134)